tests/laph_test.cpp

```diff
@@ -0,0 +1,161 @@
+#include <cstdlib>
+#include <algorithm>
+#include <vector>
+
+#include "instantiate.h"
+#include "test.h"
+
+/*
+  This test will perhaps eventually evolve into a full test for 3-d
+  Laplace eigenvector generation and sink projection.  For now the
+  focus is eigen-vector projection
+ */
+
+using test_t = std::tuple<QudaPrecision, int, int, int, int>;
+
+using ::testing::Combine;
+using ::testing::get;
+using ::testing::Values;
+
+struct LaphTest : ::testing::TestWithParam<test_t> {
+  test_t param;
+  LaphTest() : param(GetParam()) { }
+};
+
+auto laph_test(test_t param)
+{
+  using namespace quda;
+
+  QudaPrecision precision = get<0>(param);
+  int nSink = get<1>(param);
+  int nEv = get<2>(param);
+  int tileSink = get<3>(param);
+  int tileEv = get<4>(param);
+
+  constexpr int nSpin = 4;
+  constexpr int nColor = 3;
+
+  QudaInvertParam invParam = newQudaInvertParam();
+  invParam.cpu_prec = QUDA_DOUBLE_PRECISION;
+  invParam.cuda_prec = precision;
+  invParam.gamma_basis = QUDA_DEGRAND_ROSSI_GAMMA_BASIS;
+  invParam.dirac_order = QUDA_DIRAC_ORDER;
+
+  ColorSpinorParam cs_param;
+  cs_param.nColor = nColor;
+  cs_param.nSpin = nSpin;
+  cs_param.x = {xdim, ydim, zdim, tdim};
+  cs_param.siteSubset = QUDA_FULL_SITE_SUBSET;
+  cs_param.setPrecision(invParam.cpu_prec);
+  cs_param.siteOrder = QUDA_EVEN_ODD_SITE_ORDER;
+  cs_param.fieldOrder = QUDA_SPACE_SPIN_COLOR_FIELD_ORDER;
+  cs_param.gammaBasis = invParam.gamma_basis;
+  cs_param.pc_type = QUDA_4D_PC;
+  cs_param.location = QUDA_CPU_FIELD_LOCATION;
+  cs_param.create = QUDA_NULL_FIELD_CREATE;
+
+  ColorSpinorField rngDummy(cs_param);
+  RNG rng(rngDummy, 1234);
+
+  // initialize quark sinks
+  std::vector<ColorSpinorField> sinkList(nSink);
+  for (auto &s : sinkList) {
+    s = ColorSpinorField(cs_param);
+    spinorNoise(s, rng, QUDA_NOISE_GAUSS);
+  }
+
+  // initialize EVs
+  cs_param.nSpin = 1;
+  std::vector<ColorSpinorField> evList(nEv);
+  for (auto &e : evList) {
+    e = ColorSpinorField(cs_param);
+    spinorNoise(e, rng, QUDA_NOISE_GAUSS);
+  }
+
+  // host reference [nSink][nEv][Lt][nSpin][complexity]
```

Comment:
I don't necessarily like how this host verification is inlined within the test file as opposed to being placed in a separate file in the host reference directory, but it's not a hill I'm going to die on.